### ISSUE TITLE
Error in variables[[j]] : subscript out of bounds

### ISSUE THREAD

**Issue by jjnesbitt** [KitwareMedical/UTM#1]
This errors seems to occur in https://github.com/KitwareMedical/UTM/blob/main/Scripts/Analysis/analyse.images.cor.R
Full error message:
```
Error in variables[[j]] : subscript out of bounds
Calls: analysis.correlations
Execution halted
```

This occurs after calling https://github.com/KitwareMedical/UTM/blob/main/Scripts/run.utm.barycenter.R, from [this](https://github.com/girder/otm-server/blob/e759f9c3c066795ea8319702333adfe6cd11c580/optimal_transport_morphometry/core/tasks.py#L169-L178) subprocess call in https://github.com/girder/otm-server

**Comment by ebrahimebrahim** [comment]
The `variables` object is loaded from `variables.Rdata`, as defined [here](https://github.com/KitwareMedical/UTM/blob/ba89124c8964410f959ecf9085cb928d74c591ee/Scripts/get.configuration.R#L8). The Rdata file is saved [here](https://github.com/KitwareMedical/UTM/blob/ba89124c8964410f959ecf9085cb928d74c591ee/Scripts/Processing/setup.R#L98), using data loaded from csv file [here](https://github.com/KitwareMedical/UTM/blob/ba89124c8964410f959ecf9085cb928d74c591ee/Scripts/Processing/setup.R#L20). The csv file is the one passed [here](https://github.com/girder/otm-server/blob/e759f9c3c066795ea8319702333adfe6cd11c580/optimal_transport_morphometry/core/tasks.py#L174) as a command line argument to `run.utm.barycenter.R`.

This might help figure out what's going on. E.g. is the variables Rdata file getting created? What's in it?

I can also look more into this once I can reproduce the error by running otm-server myself.

**Comment by jjnesbitt** [comment]
Thanks for that, past the CSV file itself I was unaware of the workflow.

The CSV file is being created, and contains just the `name' column (shown [here](https://github.com/girder/otm-server/blob/e759f9c3c066795ea8319702333adfe6cd11c580/optimal_transport_morphometry/core/tasks.py#L143-L144)). That is, the entire csv file is just:

```
name
OAS1_0001_MR1.nii
OAS1_0002_MR1.nii
OAS1_0003_MR1.nii
OAS1_0004_MR1.nii
OAS1_0005_MR1.nii
OAS1_0006_MR1.nii
OAS1_0007_MR1.nii
OAS1_0009_MR1.nii
OAS1_0010_MR1.nii
OAS1_0011_MR1.nii

```

**Comment by ebrahimebrahim** [comment]
Hmm if it's just the name column then there are no features with which to create a correlation map, so maybe analyse.images.cor.R is getting confused? There should be some other columns as seen in [the sample here](https://github.com/girder/otm-server/blob/master/sample_data/oasis_small.csv).

**Comment by jjnesbitt** [comment]
> Hmm if it's just the name column then there are no features with which to create a correlation map, so maybe analyse.images.cor.R is getting confused? There should be some other columns as seen in [the sample here](https://github.com/girder/otm-server/blob/master/sample_data/oasis_small.csv?rgh-link-date=2022-04-19T21%3A30%3A01Z).

Interesting. The code that I started with only ever pulled out the `name` field, so maybe that needs to be updated.

**Comment by ebrahimebrahim** [comment]
> The code that I started with only ever pulled out the `name` field, so maybe that needs to be updated.

Can you point out where you saw this?

**Comment by jjnesbitt** [comment]
> Can you point out where you saw this?

[Here](https://github.com/girder/otm-server/blob/e759f9c3c066795ea8319702333adfe6cd11c580/optimal_transport_morphometry/core/tasks.py#L140-L155) is the block of code that creates the `variables` list and writes it out.

**Comment by ebrahimebrahim** [comment]
> only ever pulled out the `name` field

The `name` key is added explicitly to `meta`, but `meta` is defined [here](https://github.com/girder/otm-server/blob/e759f9c3c066795ea8319702333adfe6cd11c580/optimal_transport_morphometry/core/tasks.py#L142) so it starts out with any metadata that was attached to the `Image`.

I'm not sure where `Image`s are created. Maybe [here](https://github.com/girder/otm-server/blob/e759f9c3c066795ea8319702333adfe6cd11c580/optimal_transport_morphometry/core/rest/image.py#L57)? If that's correct, then the metadata comes from [here](https://github.com/girder/otm-server/blob/e759f9c3c066795ea8319702333adfe6cd11c580/optimal_transport_morphometry/core/batch_parser.py#L24) in the creation of a `PendingUpload`. Here we see that `row` gets to keep whatever columns were in the original csv file, such as subject age, etc.

**Comment by jjnesbitt** [comment]
Your assessment is correct, but I was performing this from my branch which [modifies](https://github.com/girder/otm-server/blob/c8c4218865ab05f95565d9ec56894b725ce1968c/optimal_transport_morphometry/core/batch_parser.py#L36-L46) this batch processing, to create the images there. In that change, I'm just specifing `ID` when creating the Images.

**Comment by ebrahimebrahim** [comment]
Ah, then this might be what's causing the `Error in variables[[j]]`. The `analyse.images.cor.R` script is supposed to compute correlations between each voxel position and each "variable", where "variable" refers to a feature column from the input csv file. 

I'm not certain but that's a guess.

(It's a little difficult to debug the R scripts with the way they are set up. Each R script in the pipeline outputs a bunch of stuff for the next R script to ingest. There is not much data validation in between. Also there's a kind of evil "global state" problem, where the state of the program as a whole is buried in a collection of Rdata files. Fun stuff.)

**Comment by jjnesbitt** [comment]
After learning what the correct workflow for the server is, this might be moot. Although it might still be beneficial to make the app more robust to inputs like this. @ebrahimebrahim what do you think?

**Comment by ebrahimebrahim** [comment]
> After learning what the correct workflow for the server is, this might be moot. Although it might still be beneficial to make the app more robust to inputs like this. @ebrahimebrahim what do you think?

Agreed; tracking robustness improvements in #2.

**Comment by ebrahimebrahim** [comment]
For the record, since I just figured it out: the explicit cause of this error was due to the definition of the range of the [for loop here](https://github.com/KitwareMedical/UTM/blob/f6be9e66952c3bd7a82558fac33eadcfa0d78db5/Scripts/Analysis/analyse.images.cor.R#L100). In the case where `variables` is an empty list, such as when there are no feature columns in the input data table, the range becomes `1:0`. I would have have thought this is an empty vector, but no-- in R `1:0` counts backwards giving the vector `c(1,0)`. Hence the subscript out of bounds error.
